File: simple_automations/scheduling/python/cron_schedule.py
import os
import time
from pathlib import Path

import schedule

old_files_folder_name = "old_files"

def clean_up_downloads():
    print("Cleaning up Downloads")

    # get all items from the downloads filder
    download_folder_path = os.path.join(Path.home(), "Downloads", "Downloads")
    download_items = os.listdir(download_folder_path)
    moved_items = 0

    # create the old files folder if not present
    old_files_folder_path = os.path.join(download_folder_path, old_files_folder_name)
    if old_files_folder_name not in download_items:
        print(f"No {old_files_folder_name} folder yet, creating folder")
        os.mkdir(old_files_folder_path)

    # create new folder with todays timestamp
    timestamp = time.strftime("%Y_%m_%d")
    datetime_folder_path = os.path.join(old_files_folder_path, timestamp)
    if not os.path.exists(datetime_folder_path):
        print(f"No {datetime_folder_path} folder yet, creating folder")
        os.mkdir(datetime_folder_path)
    else:
        print(f"{timestamp} folder already exists in {old_files_folder_name}")

    # rename all items to move them into the current datetime folder
    to_be_moved = [item for item in download_items if item != old_files_folder_name]

    for item in to_be_moved:
        print(f"Moving {item} to {datetime_folder_path} folder")
        old_path = os.path.join(download_folder_path, item)
        new_path = os.path.join(datetime_folder_path, item)
        os.rename(old_path, new_path)

        moved_items += 1

    print(f"Moved {moved_items} of {len(to_be_moved)} items")

# clean up the downloads folder every monday
schedule.every().monday.do(clean_up_downloads)

# keep the script running and sleep in between the checks
while True:
    schedule.run_pending()
    # sleep 24h
    time.sleep(60 * 60 * 24)

File: simple_automations/scheduling/javascript/cron_schedule.js
// npm install node-schedule
const cron = require('node-schedule');
const os = require('os');
const path =  require('path');
const fs = require('fs');

const oldFilesFolderName = 'old_files';
const downloadsFolder = path.join(os.homedir(), 'Downloads', 'test');
const oldFilesFolderPath = path.join(downloadsFolder, oldFilesFolderName);

function cleanUpDownloads() {
  console.log('Starting Cleanup');
  // create old files folder
  if (!fs.existsSync(oldFilesFolderPath)) {
    fs.mkdirSync(oldFilesFolderPath);
    console.log('Created old files folder');
  }

  const now = new Date();
  // months start at 0 so add 1
  const timestamp = `${now.getFullYear()}_${now.getMonth() + 1}_${now.getDate()}`;
  
  // create new folder with todays timestamp
  const dateFolder = path.join(oldFilesFolderPath, timestamp);
  if (!fs.existsSync(dateFolder)) {
    fs.mkdirSync(dateFolder);
    console.log('Created Date Folder.');
  }
  
  // get all items from downloads folder
  const dir = fs.readdirSync(downloadsFolder);
  
  let moved = 0;
  // move items to new folder
  for (const item of dir.filter(f => !f.includes(oldFilesFolderName))) {
    fs.renameSync(path.join(downloadsFolder, item), path.join(dateFolder, item));
    moved++;
  }

  console.log(`Cleaned ${moved} files.`);
}

// uncomment for testing
//cleanUpDownloads();

// every monday at 12 o clock
cron.scheduleJob('0 12 * * MON', cleanUpDownloads);